### PR TITLE
drivers: pwm: stm32: add pwm capture support

### LINKED ISSUE #39394: drivers:pwm:stm32: add pwm capture suport
**Is your enhancement proposal related to a problem? Please describe.**
pwm capture is usefull when handing PPM signals

### PR DESCRIPTION
**Pull request by TilmannUnte**
Extends STM32 PWM driver to support capturing pulse width, period,
or both.

Fixes #39394

Signed-off-by: Tilmann Unte <unte@es-augsburg.de>

### CODE REVIEW

**Review by gmarull — DISMISSED** [review]
Thanks for this addition, however, this PR seems to mix the changes with a few code style tweaks, making the review process difficult. Code style changes are ok as far as they are contained in their own commit, but not mixed with a feature addition. The overlay file for l475 should also be part of a separate commit.

**Review by ABOSTM — CHANGES_REQUESTED** [review]
@TilmannUnte, Thanks for this PR.
This is a nice job and this feature was awaited.
Some inline comments

**Comment by TilmannUnte** [comment]
> this link https://github.com/multiwii/baseflight/blob/master/src/drv_pwm.c maybe helpful

I had not seen this before. Unfortunately it does not help much as a reference, because it uses the HAL library, whereas the Zephyr driver exclusively uses LL.

I personally prefer LL, because it has a close relationship to the hardware registers defined in the reference manuals.

File: drivers/pwm/pwm_stm32.c [hunk 1]
```diff
@@ -28,10 +28,25 @@ LOG_MODULE_REGISTER(pwm_stm32, CONFIG_PWM_LOG_LEVEL);
...
+	pwm_capture_callback_handler_t callback;
+	void *user_data;
+	uint32_t period;
+	uint32_t pulse;
+	uint32_t overflows;
+	uint8_t skip_irq;
+	bool capture_period;
+	bool capture_pulse;
+	bool continuous;
+};
```
[the remark this answers is not in the document]
Comment (TilmannUnte) [reply drivers/pwm/pwm_stm32.c:41]: Done!

File: drivers/pwm/pwm_stm32.c [hunk 2]
```diff
@@ -259,6 +293,240 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
...
+		uint32_t pwm, pwm_flags_t flags,
+		pwm_capture_callback_handler_t cb, void *user_data)
+{
+
+	/*
+	 * From the reference manual RM0351:
+	 * The PWM input mode can be used only with the TIMx_CH1/TIMx_CH2
+	 * signals due to the fact that only TI1FP1 and TI2FP2 are connected to
+	 * the slave mode controller.
+	 */
```
[the remark this answers is not in the document]
Comment (TilmannUnte) [reply drivers/pwm/pwm_stm32.c:348]: I improved the comment for clarity following your suggestions.

File: drivers/pwm/pwm_stm32.c [hunk 3]
```diff
@@ -241,6 +271,10 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
 		oc_init.CompareValue = pulse_cycles;
 		oc_init.OCPolarity = get_polarity(flags);
 
+		LL_TIM_SetSlaveMode(cfg->timer, LL_TIM_SLAVEMODE_DISABLED);
+		LL_TIM_SetTriggerInput(cfg->timer, LL_TIM_TS_ITR0);
+		LL_TIM_DisableMasterSlaveMode(cfg->timer);
+
```
[the remark this answers is not in the document]
Comment (TilmannUnte) [reply drivers/pwm/pwm_stm32.c:277]: I decided to use both

File: drivers/pwm/pwm_stm32.c [hunk 4]
```diff
@@ -259,6 +293,243 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
...
+		LOG_ERR("No PWM capture type specified");
+		return -EINVAL;
+	}
+
+	if (!IS_TIM_SLAVE_INSTANCE(cfg->timer)) {
+		LOG_ERR("Timer does not support slave mode for PWM capture");
+		return -ENOTSUP;
+	}
+
+	cpt->callback = cb;
```
[the remark this answers is not in the document]
Comment (TilmannUnte) [reply drivers/pwm/pwm_stm32.c:376]: Yeah, seems like the better way. I also followed your other suggestions related to this one.

File: drivers/pwm/pwm_stm32.c [hunk 5]
```diff
@@ -224,6 +247,13 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
 		return -ENOTSUP;
 	}
 
+#ifdef CONFIG_PWM_CAPTURE
+	if (LL_TIM_IsEnabledIT_CC1(cfg->timer)) {
```
[the remark this answers is not in the document]
Comment (TilmannUnte) [reply drivers/pwm/pwm_stm32.c:251]: Must've somehow missed that. I also added checks for specific channel numbers.

File: drivers/pwm/pwm_stm32.c [hunk 6]
```diff
@@ -259,6 +293,243 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
...
+					cpt->capture_period ? cpt->period : 0u,
+					cpt->capture_pulse ? cpt->pulse : 0u,
+					status, cpt->user_data);
+			}
+		}
+	} else {
+		cpt->skip_irq--;
+	}
+	LL_TIM_ClearFlag_CC1(cfg->timer);
+	LL_TIM_ClearFlag_CC2(cfg->timer);
```
[the remark this answers is not in the document]
Comment (TilmannUnte) [reply drivers/pwm/pwm_stm32.c:529]: That wouldn't be good, I added the check.

File: drivers/pwm/pwm_stm32.c [hunk 7]
```diff
@@ -259,6 +293,243 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
...
+
+			if (cpt->callback) {
+				cpt->callback(dev, in_ch,
+					cpt->capture_period ? cpt->period : 0u,
+					cpt->capture_pulse ? cpt->pulse : 0u,
+					status, cpt->user_data);
+			}
+		}
+	} else {
+		cpt->skip_irq--;
```
[the remark this answers is not in the document]
Comment (TilmannUnte) [reply drivers/pwm/pwm_stm32.c:526]: I'm clearing the update flag seperately now.

skip_irq is meant to only decrement when a capture occurs. So I included that check.
The update flag only starts to have relevance once the skipped irqs are over.

**Comment by TilmannUnte** [comment]
I've worked on the requested changes, but will continue to test locally

**Review by ABOSTM — APPROVED** [review]
LGTM

**Comment by ABOSTM** [comment]
BTW, I tested on nucleo_f207zg, and it works fine.

File: tests/drivers/pwm/pwm_loopback/boards/disco_l475_iot1.overlay [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		pinctrl-0 = <&tim5_ch1_pa0>; /* CN3 D1 */
+	};
+};
+
+/* 16-Bit timers */
+&timers3 {
+	status = "okay";
+	pwm3: pwm {
+		status = "okay";
+		st,prescaler = <255>;
```
Comment (erwango) [inline tests/drivers/pwm/pwm_loopback/boards/disco_l475_iot1.overlay:40]: Should be sync'ed with https://github.com/zephyrproject-rtos/zephyr/pull/39628.
This won't trigger any failure, but we should avoid having both implementations in tree once #39628 is merged

File: drivers/pwm/pwm_stm32.c [hunk 9]
```diff
@@ -331,9 +625,28 @@ static int pwm_stm32_init(const struct device *dev)
...
+#define IRQ_CONFIG_FUNC(index)                                                 \
+static void pwm_stm32_irq_config_func_##index(const struct device *dev)        \
+{                                                                              \
+	IRQ_CONNECT(DT_IRQN(DT_PARENT(DT_DRV_INST(index))),                    \
+			DT_IRQ(DT_PARENT(DT_DRV_INST(index)), priority),       \
+	pwm_stm32_isr, DEVICE_DT_INST_GET(index), 0);                          \
+	irq_enable(DT_IRQN(DT_PARENT(DT_DRV_INST(index))));                    \
+}
+#define CAPTURE_INIT(index)                                                    \
+.irq_config_func = pwm_stm32_irq_config_func_##index
```
Comment (erwango) [inline drivers/pwm/pwm_stm32.c:644]: Nitpick: indentation

File: drivers/pwm/pwm_stm32.c [hunk 10]
```diff
@@ -331,9 +625,28 @@ static int pwm_stm32_init(const struct device *dev)
...
 	return 0;
 }
 
+#ifdef CONFIG_PWM_CAPTURE
+#define IRQ_CONFIG_FUNC(index)                                                 \
+static void pwm_stm32_irq_config_func_##index(const struct device *dev)        \
+{                                                                              \
+	IRQ_CONNECT(DT_IRQN(DT_PARENT(DT_DRV_INST(index))),                    \
+			DT_IRQ(DT_PARENT(DT_DRV_INST(index)), priority),       \
+	pwm_stm32_isr, DEVICE_DT_INST_GET(index), 0);                          \
```
Comment (erwango) [inline drivers/pwm/pwm_stm32.c:640]: Nitpick: indentation

File: drivers/pwm/pwm_stm32.c [hunk 11]
```diff
@@ -331,9 +625,28 @@ static int pwm_stm32_init(const struct device *dev)
 
 	LL_TIM_EnableCounter(cfg->timer);
 
+#ifdef CONFIG_PWM_CAPTURE
+	cfg->irq_config_func(dev);
+#endif /* CONFIG_PWM_CAPTURE */
```
Comment (erwango) [inline drivers/pwm/pwm_stm32.c:630]: Break a line

File: drivers/pwm/pwm_stm32.c [hunk 12]
```diff
@@ -259,6 +303,251 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
...
+	LL_TIM_SetUpdateSource(cfg->timer, LL_TIM_UPDATESOURCE_COUNTER);
+	if (pwm == 1u) {
+		LL_TIM_EnableIT_CC1(cfg->timer);
+	} else {
+		LL_TIM_EnableIT_CC2(cfg->timer);
+	}
+	LL_TIM_EnableIT_UPDATE(cfg->timer);
+	LL_TIM_CC_EnableChannel(cfg->timer, LL_TIM_CHANNEL_CH1);
+	LL_TIM_CC_EnableChannel(cfg->timer, LL_TIM_CHANNEL_CH2);
+	return 0;
```
Comment (erwango) [inline drivers/pwm/pwm_stm32.c:459]: Break a line

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this work, this brings a long awaited feature.

A couple of nitpicks otherwise LGTM.

Main point will be to synch with https://github.com/zephyrproject-rtos/zephyr/pull/39628
(which we'll try to get merged quickly first)

File: tests/drivers/pwm/pwm_loopback/boards/disco_l475_iot1.overlay (hunk 8 above)
Comment (TilmannUnte) [reply tests/drivers/pwm/pwm_loopback/boards/disco_l475_iot1.overlay:40]: I worked out the minor nitpicks. Awaiting #39628 now, thank you for pointing it out to me.

Comment (erwango) [reply tests/drivers/pwm/pwm_loopback/boards/disco_l475_iot1.overlay:40]: @TilmannUnte https://github.com/zephyrproject-rtos/zephyr/pull/39628 has just been merged.

Comment (TilmannUnte) [reply tests/drivers/pwm/pwm_loopback/boards/disco_l475_iot1.overlay]: All done!

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
No related CI failure: tests/kernel/lifo/lifo_usage/kernel.lifo.usage in qemu_cortex_a53_smp:tests/kernel/lifo/lifo_usage/kernel.lifo

**Comment by TilmannUnte** [comment]
> No related CI failure: tests/kernel/lifo/lifo_usage/kernel.lifo.usage in qemu_cortex_a53_smp:tests/kernel/lifo/lifo_usage/kernel.lifo

Anything I can/need to do here?

**Comment by erwango** [comment]
> > No related CI failure: tests/kernel/lifo/lifo_usage/kernel.lifo.usage in qemu_cortex_a53_smp:tests/kernel/lifo/lifo_usage/kernel.lifo
> 
> Anything I can/need to do here?

You can have a try to rebase on very latest master, hopefully the issue got fixed.

**Pull request merged**